### PR TITLE
Fix line endings for public api fixers

### LINKED ISSUE #4749: PublicAPIAnalyzer CodeFix does not respect existing newline style
https://github.com/dotnet/roslyn-analyzers/blob/62f71ea5c1ad04f9c4e5579c5c2b97fab16ff967/src/PublicApiAnalyzers/Core/CodeFixes/DeclarePublicApiFix.cs#L110

Instead of reading the line endings that already exist in the file, the public API analyzers use the default from the environment. This is a problem on Windows when autocrlf is off, or when the source is shared with a Linux partition (such as WSL or when running Windows in a VM) and the line endings are set to just `\n`. Instead, it should be looking at the existing line endings in the file and, if they are just `\n`, only using that, instead of unconditionally using the environment-default newline.

### PR DESCRIPTION
**Pull request by Youssef1313**
Fixes #4749

### CODE REVIEW

File: src/PublicApiAnalyzers/Core/CodeFixes/DeclarePublicApiFix.cs [hunk 1]
```diff
@@ -136,7 +143,14 @@ private static SourceText RemoveSymbolNamesFromSourceText(SourceText sourceText,
             List<string> lines = GetLinesFromSourceText(sourceText);
             IEnumerable<string> newLines = lines.Where(line => !linesToRemove.Contains(line));
 
-            SourceText newSourceText = sourceText.Replace(new TextSpan(0, sourceText.Length), string.Join(Environment.NewLine, newLines) + GetEndOfFileText(sourceText));
+            var endOfLine = Environment.NewLine;
+            if (sourceText.Lines.Count > 1)
+            {
+                var firstLine = sourceText.Lines[0];
+                endOfLine = sourceText.ToString(new TextSpan(firstLine.End, firstLine.EndIncludingLineBreak - firstLine.End));
+            }
```
Comment (sharwell) [inline src/PublicApiAnalyzers/Core/CodeFixes/DeclarePublicApiFix.cs:151]: ❔ Can we make this an extension method? It seems to be used in several places.

**Review by sharwell — CHANGES_REQUESTED** [review]
* Concept looks good
* Can be improved by the use of an extension method
* Tests are needed

File: src/PublicApiAnalyzers/Core/CodeFixes/DeclarePublicApiFix.cs [hunk 2]
```diff
@@ -51,16 +51,6 @@ public sealed override Task RegisterCodeFixesAsync(CodeFixContext context)
...
 
-        internal static TextDocument? GetUnshippedDocument(Project project)
-        {
-            return project.AdditionalDocuments.FirstOrDefault(doc => doc.Name.Equals(DeclarePublicApiAnalyzer.UnshippedFileName, StringComparison.Ordinal));
-        }
-
-        internal static TextDocument? GetShippedDocument(Project project)
-        {
-            return project.AdditionalDocuments.FirstOrDefault(doc => doc.Name.Equals(DeclarePublicApiAnalyzer.ShippedFileName, StringComparison.Ordinal));
-        }
```
Comment (Youssef1313) [inline src/PublicApiAnalyzers/Core/CodeFixes/DeclarePublicApiFix.cs:62]: These were used by different code fixes. So I moved them to the new helper class.

**Comment by Youssef1313** [comment]
Closing and re-opening to re-run CI due to failing flaky test: https://github.com/dotnet/roslyn-analyzers/issues/4544

File: src/PublicApiAnalyzers/UnitTests/DeclarePublicApiAnalyzerTests.cs [hunk 3]
```diff
@@ -1481,6 +1481,46 @@ public class C
             await VerifyCSharpAdditionalFileFixAsync(source, shippedText, unshippedText, fixedUnshippedText);
         }
 
+        [Theory]
+        [WorkItem(4749, "https://github.com/dotnet/roslyn-analyzers/issues/4749")]
+        [InlineData("\r\n")] // Windows line ending.
+        [InlineData("\n")] // Linux line ending.
+        [InlineData("\r")] // macOS line ending.
```
Comment (sharwell) [inline src/PublicApiAnalyzers/UnitTests/DeclarePublicApiAnalyzerTests.cs:1488]: This was only the Mac line ending prior to OSX. There is no need to support it.

```suggestion
```

Comment (sharwell) [reply src/PublicApiAnalyzers/UnitTests/DeclarePublicApiAnalyzerTests.cs:1488]: Mildly interesting side note: the first release of OSX predated .NET Framework 1.0 by almost a year, so C# code should really never need to handle these.

File: src/PublicApiAnalyzers/UnitTests/DeclarePublicApiAnalyzerTests.cs [hunk 4]
```diff
@@ -2167,7 +2206,7 @@ public class {|RS0016:{|RS0016:C2|}|} { }
             await VerifyCSharpAdditionalFileFixAsync(source, shippedText, unshippedText, fixedUnshippedText);
         }
 
-        [WindowsOnlyTheory]
+        [Theory]
```
Comment (Youssef1313) [inline src/PublicApiAnalyzers/UnitTests/DeclarePublicApiAnalyzerTests.cs:2209]: I expect this test to be fixed on Linux now. Waiting for CI Linux build to confirm.

**Comment by Evangelink** [comment]
@sharwell Could you review again please?

**Review by sharwell — APPROVED** [review]

**Pull request merged**